pyClient/zeth/wallet.py

```diff
@@ -6,40 +6,186 @@
+        self.commitment = commitment
+
+    def as_input(self) -> Tuple[int, ZethNote]:
+        """
+        Returns the description in a form suitable for joinsplit.
+        """
+        return (self.address, self.note)
+
+    def to_json(self) -> str:
+        json_dict = {
+            "note": joinsplit.zeth_note_to_json_dict(self.note),
+            "address": str(self.address),
+            "commitment": self.commitment.hex(),
+        }
+        return json.dumps(json_dict, indent=4)
+
+    @staticmethod
+    def from_json(json_str: str) -> ZethNoteDescription:
+        json_dict = json.loads(json_str)
+        return ZethNoteDescription(
+            note=joinsplit.zeth_note_from_json_dict(json_dict["note"]),
+            address=int(json_dict["address"]),
+            commitment=bytes.fromhex(json_dict["commitment"]))
+
+
 class Wallet:
-    def __init__(self, username: str, wallet_dir: str, k_sk_receiver: PrivateKey):
+    """
+    Very simple class to track the list of notes owned by a Zeth user.
+
+    Note: this class does not store the notes in encrypted form, and encodes
+    some information (including value) in the filename. It is a proof of
+    concept implementation and NOT intended to be secure against intruders who
+    have access to the file system. However, we expect that a secure
+    implementation could expose similar interface and functionality.
+    """
+    def __init__(
+            self,
+            mixer_instance: Any,
+            username: str,
+            wallet_dir: str,
+            k_sk_receiver: PrivateKey):
+        assert "_" not in username
+        self.mixer_instance = mixer_instance
         self.username = username
         self.wallet_dir = wallet_dir
         self.k_sk_receiver = k_sk_receiver
         self.k_sk_receiver_bytes = \
             k_sk_receiver.encode(encoder=encoding.RawEncoder)
+        self.state_file = join(wallet_dir, f"state_{username}")
+        _ensure_dir(self.wallet_dir)
 
     def receive_notes(
             self,
-            addrs_and_ciphertexts: List[Tuple[int, bytes]],
-            k_pk_sender: PublicKey) -> List[Tuple[int, ZethNote]]:
-        new_notes_iter = joinsplit.receive_notes(
-            addrs_and_ciphertexts, k_pk_sender, self.k_sk_receiver)
+            encrypted_notes: List[EncryptedNote],
+            k_pk_sender: PublicKey) -> List[ZethNoteDescription]:
+        """
+        Decrypt any notes we can, verify them as being valid, and store them in
+        the database.
+        """
         new_notes = []
-        for addr, note in new_notes_iter:
-            print(
-                f"[INFO] {self.username} received payment: {note} (addr: {addr})")
-            self._write_note(note)
-            new_notes.append((addr, note))
+        addr_note_iter = joinsplit.receive_notes(
+            encrypted_notes, k_pk_sender, self.k_sk_receiver)
+        for addr, note in addr_note_iter:
+            note_desc = self._check_note(addr, note)
```

Comment:
@dtebbs I think this code is leaking a ton of information about the recipient since it calls the `_check_note` function which itself call the `get_merkle_leaf` function if the decryption is successful (ie: if you are the intended recipient of a payment).
Can you please confirm (or challenge if I misunderstand the code) ?

Reply:
We no longer perform a query here (see discussion about commitment events)